src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php

```diff
@@ -162,56 +164,140 @@ private function compileRoutes(RouteCollection $routes, $supportsRedirections)
+        $staticRoutes = array();
+        $dynamicRoutes = new RouteCollection();
+        $dynamicRegex = array();
+        $code = '';
 
-        foreach ($collection as $dumperRoute) {
-            $prefix = $dumperRoute->getRoute()->compile()->getStaticPrefix();
-            $prefixCollection->addRoute($prefix, $dumperRoute);
-        }
+        foreach ($collection->all() as $name => $route) {
+            $compiledRoute = $route->compile();
+            $regex = $compiledRoute->getRegex();
+            $methods = $route->getMethods();
+            if ($hasTrailingSlash = $supportsRedirections && $pos = strpos($regex, '/$')) {
+                $regex = substr_replace($regex, '/?$', $pos, 2);
+            }
+            if (!$compiledRoute->getPathVariables()) {
+                $url = $route->getPath();
+                if ($hasTrailingSlash) {
+                    $url = rtrim($url, '/');
+                }
+                foreach ($dynamicRegex as $rx) {
+                    if (preg_match($rx, $url)) {
+                        $dynamicRegex[] = $regex;
+                        $dynamicRoutes->add($name, $route);
+                        continue 2;
+                    }
+                }
 
-        $prefixCollection->optimizeGroups();
+                $staticRoutes[$url][$name] = array($hasTrailingSlash, $route);
+            } else {
+                $dynamicRegex[] = $regex;
+                $dynamicRoutes->add($name, $route);
+            }
+        }
 
-        return $prefixCollection;
+        return array($staticRoutes, $dynamicRoutes);
     }
 
-    /**
-     * Generates PHP code to match a tree of routes.
-     *
-     * @param StaticPrefixCollection $collection           A StaticPrefixCollection instance
-     * @param bool                   $supportsRedirections Whether redirections are supported by the base class
-     * @param string                 $ifOrElseIf           either "if" or "elseif" to influence chaining
-     *
-     * @return string PHP code
-     */
-    private function compileStaticPrefixRoutes(StaticPrefixCollection $collection, $supportsRedirections, $ifOrElseIf = 'if')
+    private function compileStaticRoutes(array $staticRoutes, bool $supportsRedirections): string
     {
+        if (!$staticRoutes) {
+            return '';
+        }
         $code = '';
-        $prefix = $collection->getPrefix();
 
-        if (!empty($prefix) && '/' !== $prefix) {
-            $code .= sprintf("    %s (0 === strpos(\$pathinfo, %s)) {\n", $ifOrElseIf, var_export($prefix, true));
+        foreach ($staticRoutes as $url => $routes) {
+            $code .= sprintf("        case %s:\n", self::export($url));
+            foreach ($routes as $name => list($hasTrailingSlash, $route)) {
+                $methods = $route->getMethods();
+                $supportsTrailingSlash = $supportsRedirections && (!$methods || in_array('HEAD', $methods) || in_array('GET', $methods));
```

Comment:
Route methods must be [uppercase strings](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Routing/Route.php#L225), `\in_array` calls could be made strict

Reply:
this is borrowed from existing code, nothing to do here: uppercase is already done by the very method your linked, and tweaking in_array() would have zero impact as this is dumping code, not critical at all